### ISSUE TITLE
Result object is memory hungry (in place sorting?)

### ISSUE THREAD

**Issue by gamesguru** [farzher/fuzzysort#56]
We are using this to search millions of product descriptions but it returns a large result object foreach product.

If we could do in placr sorting or have it return a simple order int array, it would consume less memory but still allow us to sort results. 

Thanks so much, we love the accuracy and speed of the library

**Comment by gamesguru** [comment]
@farzher is this something supportrd out of the box, or can be with little tweaking?

**Comment by farzher** [comment]
i don't think sorting is the issue.  it's a good algo (heapsort) and if you use `limit: 100` it'll only sort/return 100 things.

it caching all target data to makes searches faster is what takes up so much memory. (those large result objects are all from that cache, results only reference the cache)

to be clear, there's no issue? besides you notice it uses a lot of memory? it's not slow?

**Comment by gamesguru** [comment]
well if sorts in place that might switch around our original list of product names?

I guess we would want it to return something like a list of high scores paired with the original index in product name array. that would make the result take up less memory.

I don't think it was that slow, we can do some more benchmarks against less flexible search algorithms and report shortly

**Comment by gamesguru** [comment]
it takes about 100-200ms for searching 1.1 million product names.  that is pretty fast.

but i would like it to not make copies of the original list.  it exceeds 1gb of memory with one search, and we need to support up to 10-20 people searching the server at once!  is there any trade-off or way to accomplish this?

i dont even need it sorted.  just need the top 100 or 1000 scores with an index mapping them back to an entry in product name array

**Comment by farzher** [comment]
it's not currently designed to use low memory. you can try editing fuzzysort.js source code. find and replace `999` with `999||1`. it won't make a cache anymore.

**Comment by gamesguru** [comment]
there appear to be other places where the targets are being passed by value (ie. copied into memory)

```
          // no keys
        } else {
          for (var i = targetsLen - 1; i >= 0; --i) {
            var target = targets[i]  // HERE
            if (!target) continue
            if (!isObj(target)) target = fuzzysort.getPrepared(target)

            var result = algorithm(search, target, searchLowerCode)  // HERE
            if (result === null) continue
            if (result.score < threshold) continue
            if (resultsLen < limit) { q.add(result); ++resultsLen }
            else {
              ++limitedCount
              if (result.score > q.peek().score) q.replaceTop(result)
            }
          }
        }
```

**Comment by gamesguru** [comment]
i changed to `999 || 1` but it still using the same amount of memory.

after reading the product names into node, it uses 130mb.  after running `fuzzysort.go()` it takes up 530mb

**Comment by gamesguru** [comment]
I think a full solution would involve adding an index to the original targets array as well as caching the lower codes

and not sorting server side to leave all collections all modified

I can PR those on a branch soon and we can maybe add an option to use low memory in place. I don't think most people are using your library with as much data as us, but it can definitely handle it

**Comment by farzher** [comment]
probably fixed in v2.0.0
